src/s_inter.c

```diff
@@ -637,10 +642,32 @@ static int poll_fds()
             }
             else
             {
+                if(kFdpManagerIoThreadHiPriority == fp->fdp_manager)
+                    fp->fdp_manager = kFdpManagerIoThread;
                 int fd = fp->fdp_fd;
                 t_rbskt* rbskt = fp->fdp_rbskt;
                 ring_buffer* rb = rbskt->rs_rb;
                 unsigned int size = rb_available_to_write(rb);
+                if(0 == size
+                    || (rbskt->rs_preserve_boundaries
+                        && socket_bytes_available(fd) > size))
+                {
+                    // not enough space in the ring buffer: let's postpone
+                    // recv() while the audio thread empties the ringbuffer
+                    fprintf(stderr, "Throttling read on fd %d\n", fd);
+                    // by raising the priority of the fd, we ensure that this
+                    // socket doesn't "starve" for a long time (e.g.: if
+                    // other sockets would be continuously keeping the ring
+                    // buffer almost full).
+
+                    // no need for atomic store: we already "acquired" it
+                    // in fdp_select() and this is a flag for this same
+                    // thread
+                    fp->fdp_manager = kFdpManagerIoThreadHiPriority;
+                    // return 0 to encourage the calling thread to sleep
+                    // before calling again
+                    return 0;
```

Comment:
Why do you need to return here? Just do a `continue` keep going with the for loop. This would also make the `kFdpManagerIoThreadHiPriority` unnecessary.

Reply:
The buffer has little space available, so  I want to increase the available space to be able to copy to it this big incoming packet. If I don't give this packet a special priority, then there is the (remote) risk that there will never be enough space in the buffer for it.
Think many fds receiving a lot of data. If this `fd` is towards the end of the list, we may always get here and find that there is never enough space for this big packet. By giving it a higher priority I ensure that I will try this first next time (hopefully after the audio thread has freed up some space in the buffer). It may be that the audio thread is empties the buffer shortly after this `fd` is processed (and skipped). If we `continue`, the other `fds` will see the buffer almost empty and will dump all their messages in there, possibly preventing this `fd` from ever seeing enough available space. If the two threads happen to randomly synchronise like this (audio emptying the buffer just after this `fd`) for a few calls, then this packet could be delayed significantly. By increasing priority and not giving a chance to the other `fd`s to fill up the buffer, we will not starve this `fd`. The downside is that we may end up slowing the processing of all the others for a few executions.

Reply:
>  If we continue, the other fds will see the buffer almost empty and will dump all their messages in there, possibly preventing this fd from ever seeing enough available space.

I don't understand. Each fd gets its own ring buffer. If one fd skips the `recv`, this doesn't affect other fds at all. If you see that a fd can't write its data into its ringbuffer (because there is too little space), just skip it and `continue` with the next fd. Or am I completely missing something here?

Reply:
You're so right. I was confused with what happens on the output side (where there is just one rb shared by everyone. Apologies, I guess I've been in front of this code too long today!

Reply:
BTW, I think
`
if(0 == size || (rbskt->rs_preserve_boundaries && socket_bytes_available(fd) > size))
`
should be
`
if(0 == size
    || (rbskt->rs_preserve_boundaries
        && (socket_bytes_available(fd) + sizeof(ret)) > size))
`
to include the return value, right?

Reply:
> I was confused with what happens on the output side

I've already thought so :-)

Reply:
Looks good!